mathjax3-ts/output/chtml/Wrapper.ts

```diff
@@ -0,0 +1,604 @@
+        bold: {
+            normal: 'bold',
+            italic: 'bold-italic',
+            fraktur: 'bold-fraktur',
+            script: 'bold-script',
+            'sans-serif': 'bold-sans-serif',
+            'sans-serif-italic': 'sans-serif-bold-italic'
+        },
+        normal: {
+            bold: 'normal',
+            'bold-italic': 'italic',
+            'bold-fraktur': 'fraktur',
+            'bold-script': 'script',
+            'bold-sans-serif': 'sans-serif',
+            'sans-serif-bold-italic': 'sans-serif-italic'
+        }
+    };
+    public static ITALICVARIANTS: {[name: string]: StringMap} = {
+        italic: {
+            normal: 'italic',
+            bold: 'bold-italic',
+            'sans-serif': 'sans-serif-italic',
+            'bold-sans-serif': 'sans-serif-bold-italic'
+        },
+        normal: {
+            italic: 'normal',
+            'bold-italic': 'bold',
+            'sans-serif-italic': 'sans-serif',
+            'sans-serif-bold-italic': 'bold-sans-serif'
+        }
+    };
+
+    /*
+     * The factory used to create more CHTMLWrappers
+     */
+    protected factory: CHTMLWrapperFactory;
+
+    /*
+     * The parent and children of this node
+     */
+    public parent: CHTMLWrapper = null;
+    public childNodes: CHTMLWrapper[];
+
+    /*
+     * The HTML element generated for this wrapped node
+     */
+    public chtml: HTMLElement = null;
+
+    /*
+     * Styles that must be handled directly by CHTML (mostly having to do with fonts)
+     */
+    protected removedStyles: StringMap = null;
+
+    /*
+     * The explicit styles set by the node
+     */
+    protected styles: CSSStyleDeclaration = null;
+
+    /*
+     * The mathvariant for this node
+     */
+    public variant: string = '';
+
+    /*
+     * The bounding box for this node, and whether it has been computed yet
+     */
+    public bbox: BBox;
+    protected bboxComputed: boolean = false;
+
+    /*
+     * Easy access to the font parameters
+     */
+    public TeX = TexFontParams;
+
+    /*
+     * Easy access to the CHTML output jax for this node
+     */
+    get CHTML() {
+        return this.factory.chtml;
+    }
+
+    /*
+     * Easy access to the metric data for this node
+     */
+    get metrics() {
+        return this.factory.chtml.math.metrics;
+    }
+
+    /*******************************************************************/
+
+    /*
+     * @override
+     */
+    constructor(factory: CHTMLWrapperFactory, node: MmlNode, parent: CHTMLWrapper = null) {
+        super(factory, node);
+        this.parent = parent;
+        this.bbox = BBox.zero();
+        this.getStyles();
+        this.getVariant();
+        this.getScale();
+        this.childNodes = node.childNodes.map((child: Node) => {
+            return this.wrap(child as MmlNode);
+        });
+    }
+
+    /*
+     * @param{MmlNode} node  The node to the wrapped
+     * @param{CHTMLWrapper} parent  The parent wrapped node
```

Comment:
I am not sure what this is. Is this the wrapped parent node of node? Why do we have to supply it explicitly then?

Reply:
Yes, it is the wrapped parent node.  Because there is no connection from the node to the wrapper (just from the wrapper to the node), we need to store a pointer to the wrapped parent node in order to be able to access the wrapper for the parent of node.

Originally, wrapping was going to be done at the time a wrapped node was needed, but now that the entire tree is wrapped in the `constructor()`, I suppose the parents could be set at that point rather than here.  Would you prefer that?